### PR TITLE
Ear clipping triangulation and robust 2d segment intersection test

### PR DESCRIPTION
**Pull request by hankstag**
A triangulation algorithm based on ear clipping and a robust 2d segments intersection test

#### Check all that apply (change to `[x]`)
- [x] All changes meet [libigl style-guidelines](https://libigl.github.io/style-guidelines/).
- [x] Adds new .cpp file.
- [x] Adds corresponding unit test.
- [ ] Adds corresponding python binding.
- [ ] This is a minor change.

### CODE REVIEW

File: include/igl/predicates/ear_clipping.cpp [hunk 1]
```diff
@@ -0,0 +1,113 @@
+
+#include <igl/slice.h>
+#include "ear_clipping.h"
+#include "point_inside_polygon.h"
+#include "predicates.h"
+
+template <typename DerivedP>
+IGL_INLINE bool igl::predicates::is_ear(
+  const Eigen::MatrixBase<DerivedP>& P,
+  const Eigen::VectorXi& RT,
```
Comment (jiangzhongshi) [inline include/igl/predicates/ear_clipping.cpp]: template with MatrixBase<DerivedRT>
```suggestion
  const Eigen::MatrixBase<DerivedRT>& RT,
```

File: include/igl/predicates/ear_clipping.h [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+#ifndef EAR_CLIPPING
+#define EAR_CLIPPING
+
+#include <Eigen/Core>
+#include "../igl_inline.h"
+
+namespace igl
+{
+  namespace predicates
+  {
```
Comment (jiangzhongshi) [inline include/igl/predicates/ear_clipping.h]: Add documentation on input arguments and return values

File: include/igl/predicates/point_inside_polygon.h [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  namespace predicates
+  {
+      // check whether 2d point lies inside 2d polygon
+      // Inputs:
+      //   P: n*2 polygon, n >= 3
+      //   q: 2d query point
+      // Returns true if point is inside polygon
+      template <typename Scalar>
+      bool point_inside_polygon(
+          const Eigen::Matrix<Scalar,-1,2>& P,
```
Comment (jiangzhongshi) [inline include/igl/predicates/point_inside_polygon.h]: use `Eigen::MatrixBase<DerivedP>` so the function will be more general (can take arguments like `Eigen::MatrixXd`)

File: include/igl/predicates/segment_segment_intersect.h [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+  
+    // Given two segments in 2d test whether they intersect each other
+    // using orient2D
+    // 
+    // Inputs:
+    //   A:   1st endpoint of segment 1
+    //   B:   2st endpoint of segment 1
+    //   C:   1st endpoint of segment 2
+    //   D:   2st endpoint of segment 2
+    //   eps: precision used for colinear case
```
Comment (jiangzhongshi) [inline include/igl/predicates/segment_segment_intersect.h]:
```suggestion
```

File: include/igl/predicates/ear_clipping.h [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+  }
+}
+
+#ifndef IGL_STATIC_LIBRARY
+#  include "ear_clipping.cpp"
+#endif
+
+
+#endif
```
Comment (teseoch) [inline include/igl/predicates/ear_clipping.h]: add a new line at the end of the file

File: include/igl/predicates/point_inside_polygon.cpp [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+// Copyright (C) 2019 Hanxiao Shen <hanxiao@cs.nyu.edu>
+//
+// This Source Code Form is subject to the terms of the Mozilla Public License
+// v. 2.0. If a copy of the MPL was not distributed with this file, You can
+// obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "point_inside_polygon.h"
+
+template <typename DerivedP, typename DerivedQ>
+IGL_INLINE bool igl::predicates::point_inside_polygon(
```
Comment (teseoch) [inline include/igl/predicates/point_inside_polygon.cpp]: this works only for convex polygons, I would change the name to make it clear

File: include/igl/predicates/point_inside_polygon.cpp [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+    if(r == igl::predicates::Orientation::COLLINEAR || 
+       r == igl::predicates::Orientation::NEGATIVE)
+      return false;
+  }
+  return true;
+}
+
+#ifdef IGL_STATIC_LIBRARY
+template bool igl::predicates::point_inside_polygon<Eigen::Matrix<double, -1, 2, 0, -1, 2>, Eigen::Matrix<double, 1, 2, 1, 1, 2> >(Eigen::MatrixBase<Eigen::Matrix<double, -1, 2, 0, -1, 2> > const&, Eigen::MatrixBase<Eigen::Matrix<double, 1, 2, 1, 1, 2> > const&);
+#endif
```
Comment (teseoch) [inline include/igl/predicates/point_inside_polygon.cpp]: here too

File: include/igl/predicates/point_inside_polygon.cpp [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+// obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "point_inside_polygon.h"
+
+template <typename DerivedP, typename DerivedQ>
+IGL_INLINE bool igl::predicates::point_inside_polygon(
+  const Eigen::MatrixBase<DerivedP>& P,
+  const Eigen::MatrixBase<DerivedQ>& q
+){
+  for(int i=0;i<P.rows();i++){
```
Comment (teseoch) [inline include/igl/predicates/point_inside_polygon.cpp]: add assertions regarding the sizes of your input

File: include/igl/predicates/point_inside_polygon.cpp [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+    int i_1 = (i+1) % P.rows();
+    Eigen::RowVector2d a = P.row(i);
```
Comment (teseoch) [inline include/igl/predicates/point_inside_polygon.cpp]: evil, your input is a DerivedP and you are casting it to double

File: include/igl/predicates/ear_clipping.cpp [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+  
+  // check whether vertex i is an ear
+  auto is_ear = [](
+    const Eigen::MatrixBase<DerivedP>& P,
+    const Eigen::MatrixBase<DerivedRT>& RT,
+    const Eigen::VectorXi& L,
+    const Eigen::VectorXi& R,
+    const int i
+  ){
+    int a = L(i), b = R(i);
```
Comment (teseoch) [inline include/igl/predicates/ear_clipping.cpp]: everywhere you are using D and I, you should use the correct type coming from Derived

File: include/igl/predicates/ear_clipping.cpp [hunk 11]
```diff
@@ -13,42 +20,48 @@ IGL_INLINE void igl::predicates::ear_clipping(
...
+                "index type should be consistent");
   
   // check whether vertex i is an ear
   auto is_ear = [](
     const Eigen::MatrixBase<DerivedP>& P,
     const Eigen::MatrixBase<DerivedRT>& RT,
-    const Eigen::VectorXi& L,
-    const Eigen::VectorXi& R,
-    const int i
+    const Eigen::Matrix<Index,-1,1>& L,
```
Comment (jdumas) [inline include/igl/predicates/ear_clipping.cpp]: Technically, `-1` should be `Eigen::Dynamic` here.

Comment (hankstag) [reply include/igl/predicates/ear_clipping.cpp]: Yes, i agree

File: include/igl/predicates/ear_clipping.cpp [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+    if(RT(i) != 0 || RT(a) != 0 || RT(b) != 0) return false;
+    Eigen::Matrix<Scalar,1,2> pa = P.row(a);
+    Eigen::Matrix<Scalar,1,2> pb = P.row(b);
+    Eigen::Matrix<Scalar,1,2> pi = P.row(i);
+    auto r = igl::predicates::orient2d(pa,pi,pb);
+    if(r == igl::predicates::Orientation::NEGATIVE || 
+       r == igl::predicates::Orientation::COLLINEAR) return false;
+    
+    // check if any vertex is lying inside triangle (a,b,i);
+    int k=R(b);
```
Comment (teseoch) [inline include/igl/predicates/ear_clipping.cpp]: Index

File: include/igl/predicates/ear_clipping.cpp [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+
+  Eigen::Matrix<Index,Eigen::Dynamic,1> L(P.rows());
+  Eigen::Matrix<Index,Eigen::Dynamic,1> R(P.rows());
+  for(int i=0;i<P.rows();i++){
+    L(i) = Index((i-1+P.rows())%P.rows());
+    R(i) = Index((i+1)%P.rows());
+  }
+
+  Eigen::VectorXi ears; // mark ears
+  Eigen::VectorXi X; // clipped vertices
```
Comment (teseoch) [inline include/igl/predicates/ear_clipping.cpp]: index for both

File: include/igl/predicates/ear_clipping.h [hunk 14]
```diff
@@ -0,0 +1,51 @@
+// This file is part of libigl, a simple c++ geometry processing library.
+//
+// Copyright (C) 2019 Hanxiao Shen <hanxiao@cs.nyu.edu>
+//
+// This Source Code Form is subject to the terms of the Mozilla Public License
+// v. 2.0. If a copy of the MPL was not distributed with this file, You can
+// obtain one at http://mozilla.org/MPL/2.0/.
+
+#ifndef EAR_CLIPPING
```
Comment (jiangzhongshi) [inline include/igl/predicates/ear_clipping.h]:
```suggestion
#ifndef IGL_PREDICATES_EAR_CLIPPING_H
```

File: include/igl/predicates/ear_clipping.h [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+#define EAR_CLIPPING
```
Comment (jiangzhongshi) [inline include/igl/predicates/ear_clipping.h]:
```suggestion
#define IGL_PREDICATES_EAR_CLIPPING_H
```

File: include/igl/predicates/point_inside_convex_polygon.h [hunk 16]
```diff
@@ -0,0 +1,39 @@
+// This file is part of libigl, a simple c++ geometry processing library.
+//
+// Copyright (C) 2019 Hanxiao Shen <hanxiao@cs.nyu.edu>
+//
+// This Source Code Form is subject to the terms of the Mozilla Public License
+// v. 2.0. If a copy of the MPL was not distributed with this file, You can
+// obtain one at http://mozilla.org/MPL/2.0/.
+
+#ifndef PREDICATE_POINT_INSIDE_CONVEX_POLYGON_H
```
Comment (jiangzhongshi) [inline include/igl/predicates/point_inside_convex_polygon.h]:
```suggestion
#ifndef IGL_PREDICATES_POINT_INSIDE_CONVEX_POLYGON_H
```

File: include/igl/predicates/point_inside_convex_polygon.h [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+#define PREDICATE_POINT_INSIDE_CONVEX_POLYGON_H
```
Comment (jiangzhongshi) [inline include/igl/predicates/point_inside_convex_polygon.h]:
```suggestion
#define IGL_PREDICATES_POINT_INSIDE_CONVEX_POLYGON_H
```

File: include/igl/predicates/point_inside_convex_polygon.h [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+
+#include "../igl_inline.h"
+#include <Eigen/Core>
+#include "predicates.h"
+
+namespace igl
+{
+  namespace predicates
+  {
+    // check whether 2d point lies inside 2d polygon
```
Comment (jiangzhongshi) [inline include/igl/predicates/point_inside_convex_polygon.h]:
```suggestion
    // check whether 2d point lies inside 2d convex polygon
```

File: include/igl/predicates/segment_segment_intersect.h [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+// This Source Code Form is subject to the terms of the Mozilla Public License
+// v. 2.0. If a copy of the MPL was not distributed with this file, You can
+// obtain one at http://mozilla.org/MPL/2.0/.
+
+#ifndef IGL_PREDICATES_SEGMENT_SEGMENT_INTERSECT_H
+#define IGL_PREDICATES_SEGMENT_SEGMENT_INTERSECT_H
+
+#include <igl/igl_inline.h>
+#include <Eigen/Core>
+#include <igl/predicates/predicates.h>
```
Comment (jiangzhongshi) [inline include/igl/predicates/segment_segment_intersect.h]:
```suggestion
#include "predicates.h"
```

File: include/igl/predicates/point_inside_convex_polygon.h (hunk 18 above)
Comment (teseoch) [reply include/igl/predicates/point_inside_convex_polygon.h]: commit this and you are good to go!

**Pull request merged**